src/Microsoft.Azure.WebJobs.Host/Loggers/FunctionInstanceLogger.cs

```diff
@@ -17,56 +17,215 @@ internal class FunctionInstanceLogger : IFunctionInstanceLogger
     {
         private readonly ILoggerFactory _loggerFactory;
 
+        private readonly ConcurrentDictionary<string, ILogger> _loggers = new ConcurrentDictionary<string, ILogger>();
+
+        private readonly ConcurrentDictionary<string, Func<ILogger, IDictionary<string, string>, bool>> _triggerLoggers = new ConcurrentDictionary<string, Func<ILogger, IDictionary<string, string>, bool>>();
+
+        private static readonly Action<ILogger, string, string, Guid, Exception> LogFunctionStarted =
+            LoggerMessage.Define<string, string, Guid>(LogLevel.Information, 0, "Executing '{Function}' (Reason='{Reason}', Id={Id})");
+
+        private static readonly Action<ILogger, string, Guid, Exception> LogFunctionCompletedSuccess =
+            LoggerMessage.Define<string, Guid>(LogLevel.Information, 0, "Executed '{Function}' (Succeeded, Id={Id})");
+
+        private static readonly Action<ILogger, string, Guid, Exception> LogFunctionCompletedFailure =
+            LoggerMessage.Define<string, Guid>(LogLevel.Error, 0, "Executed '{Function}' (Failed, Id={Id})");
```

Comment:
These two have a different `LogLevel` and as the `Define` includes the log level in the closure, I think I see no way to do this. Is there any way for this?

Reply:
Oh you're right -- the failures are errors, which I forgot about. That kinda stinks to have almost identical messages, with just a variation in the level. Oh well, I also don't see a way around it.

Maybe one suggestion here -- should the failure and success both share the same EventId -- "FunctionCompleted"? Internally, we log this event name in our backend logs and it'd be useful to be able to write a query that groups by EventName = "FunctionCompleted" and Level -- I think that'd be really useful. Otherwise you'd have to expand the query a little to look for both events.